File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = (int)nums.size();
        for(int i=n-2;i>=0;i--){
            if(nums[i]<nums[i+1]){
                int idx=i+1;
                for(int j=i+2;j<n;j++){
                    if(nums[j]>nums[i] && nums[j]<nums[idx]){
                        idx=j;
                    }
                }
                swap(nums[idx],nums[i]);
                for(int j=idx+1;j<n;j++){
                    if(nums[j]>nums[j-1]){
                        swap(nums[j],nums[j-1]);
                    }
                }
                reverse(nums.begin()+i+1,nums.end());
                return;
            }
        }
        reverse(nums.begin(),nums.end());
    }
};

File: 31-next-permutation/next-permutation.java
class Solution {
    private void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    private void reverse(int[] arr, int sp, int ep){
        while(sp<ep){
            swap(arr, sp++, ep--);
        }
    }
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int idx = -1;
        for(int i=n-2;i>=0;i--){
            if(nums[i]<nums[i+1]){
                idx= i;
                break;
            }
        }
        if(idx!=-1){
            for(int i=n-1;i>idx;i--){
                if(nums[i]>nums[idx]){
                    swap(nums, i, idx);
                    break;
                }
            }
        }
        reverse(nums, idx+1, n-1);
    }
}